### ISSUE TITLE
Issue with LBGFS optimizer: `optimizer_.zero_grad` no longer called inside of `train_step_single`

### ISSUE THREAD

**Issue by jwdink** [skorch-dev/skorch#636]
It looks like in the following commit, the call to `optimizer_.zero_grad` was removed from `train_step_single`, and instead put inside of `train_step`:

https://github.com/skorch-dev/skorch/commit/a62e4190d38bf698a314c4283ca7a9dd7cfbe428#diff-5d1f8e13ab8b822b72e9b295cf7301c5L596

As far as I can tell, this isn't appropriate if you're using the LBFGS optimizer -- for that optimizer you need to zero the gradient on every call to the closure.

I ran into this because nets that had previously worked on skorch 0.60 now quickly hit `nan` loss on skorch 0.80. I reverted versions and hunted down this commit for the cause (see [here](https://github.com/strongio/strong-glm/blob/master/examples/ceiling_weibull.py) for an example).

**Comment by BenjaminBossan** [comment]
@jwdink Thank you a lot for reporting this error.

You are right, the commit you identified is the reason for the trouble.

This is actually a difficult problem to solve. The idea of moving the `zero_grad` out of each individual call of `optimizer.step` was to allow the user to have more control over it, e.g. to implement gradient accumulation. This is in conflict with optimizers that want to call the closure multiple times and need `zero_grad` after each of it.

As a quick fix for you, I would propose to either restore the previous state or move the call to `zero_grad` inside the `step_fn` closure here:

https://github.com/skorch-dev/skorch/blob/1f6b5428a0d58b9d0d2ef08384aeb45c1c14df55/skorch/net.py#L655-L660

I believe it should fix your issue.

Now what could be done in the long run? Hard to tell. Reverting the change removes the mentioned benefit. We could, hypothetically, add a parameter to `NeuralNet` that allows to switch between the two. But that still requires the user of LBFGS or similar optimizers to actively remember that. It's thus not a very nice solution.

If we could somehow know what optimizers actually need to call `zero_grad`, we could control the behavior dynamically. But I don't think we can know that. So this is also not a solution.

I never use optimizers that require that, thus have almost no experience. Do you have any good idea?

**Comment by jwdink** [comment]
> Now what could be done in the long run? Hard to tell. Reverting the change removes the mentioned benefit. We could, hypothetically, add a parameter to NeuralNet that allows to switch between the two. But that still requires the user of LBFGS or similar optimizers to actively remember that. It's thus not a very nice solution.

Since the change was made to accommodate additional control, would it make sense to add that argument to NeuralNet, but its default value causes the old behavior (i.e. calling `zero_grad` inside)? Then users who are doing custom things like accumulating gradients can still override it, but all the optimizers from the `torch.optim` module would work with `NeuralNet` out of the box.

> If we could somehow know what optimizers actually need to call zero_grad, we could control the behavior dynamically. But I don't think we can know that. So this is also not a solution.

I suppose one option is to just check if the optimizer is LBFGS. This wouldn't catch custom optimizers that aren't from `torch.optim`; is it too hacky to check the `__module__` and throw a warning?

**Comment by BenjaminBossan** [comment]
> but its default value causes the old behavior

Unfortunately, the changed behavior is part of our last release. Therefore, if we changed it back to the old behavior, we would create yet another breakage. But there is an argument to be made that if possible, skorch should work with the optimizers provided by PyTorch, and those include LBFGS.

So we could introduce a parameter like `perform_zero_grad_each_step=True` or so (not a very nice name) and if a user wants to do their custom stuff with gradients, they need to turn it off.

@ottonemo do you have a better idea?

**Comment by jwdink** [comment]
> So we could introduce a parameter like `perform_zero_grad_each_step=True` or so (not a very nice name) and if a user wants to do their custom stuff with gradients, they need to turn it off.

Yes sorry if I was being unclear -- this is exactly what I was proposing.

**Comment by BenjaminBossan** [comment]
No, you were not unclear, I was just elaborating on what you said :)

**Comment by BenjaminBossan** [comment]
@jwdink I created a PR (#639) to fix this issue. Could you please check if your issue is solved when you use skorch from that branch?

**Comment by jwdink** [comment]
Yep, that fixes it. Thanks for the speed response!

**Comment by BenjaminBossan** [comment]
Fixed by #639
